plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/Purity/ExpVisitor.kt

```diff
@@ -0,0 +1,134 @@
+import org.jetbrains.kotlin.formver.embeddings.expression.Old
+import org.jetbrains.kotlin.formver.embeddings.expression.PredicateAccessPermissions
+import org.jetbrains.kotlin.formver.embeddings.expression.PrimitiveFieldAccess
+import org.jetbrains.kotlin.formver.embeddings.expression.PureExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.SafeCast
+import org.jetbrains.kotlin.formver.embeddings.expression.SequentialAnd
+import org.jetbrains.kotlin.formver.embeddings.expression.SequentialOr
+import org.jetbrains.kotlin.formver.embeddings.expression.Shared
+import org.jetbrains.kotlin.formver.embeddings.expression.SharingContext
+import org.jetbrains.kotlin.formver.embeddings.expression.UnaryOperatorExpEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.UnitLit
+import org.jetbrains.kotlin.formver.embeddings.expression.While
+import org.jetbrains.kotlin.formver.embeddings.expression.WithPosition
+
+internal interface ExpVisitor<R> {
+    fun visitPureExpEmbedding(e: PureExpEmbedding): R
+    fun visitBlock(e: Block): R
+    fun visitFunctionExp(e: FunctionExp): R
+    fun visitGotoChainNode(e: GotoChainNode): R
+    fun visitIf(e: If): R
+    fun visitElvis(e: Elvis): R
+    fun visitLambdaExp(e: LambdaExp): R
+    fun visitMethodCall(e: MethodCall): R
+    fun visitSafeCast(e: SafeCast): R
+    fun visitShared(e: Shared): R
+    fun visitAssert(e: Assert): R
+    fun visitDeclare(e: Declare): R
+    fun visitEqCmp(e: EqCmp): R
+    fun visitNeCmp(e: NeCmp): R
+    fun visitBinaryOperatorExpEmbedding(e: BinaryOperatorExpEmbedding): R
+    fun visitSequentialAnd(e: SequentialAnd): R
+    fun visitSequentialOr(e: SequentialOr): R
+    fun visitInjectionBasedExpEmbedding(e: InjectionBasedExpEmbedding): R
+    fun visitFieldAccessPermissions(e: FieldAccessPermissions): R
+    fun visitForAllEmbedding(e: ForAllEmbedding): R
+    fun visitPredicateAccessPermissions(e: PredicateAccessPermissions): R
+    fun visitCast(e: Cast): R
+    fun visitIs(e: Is): R
+    fun visitOld(e: Old): R
+    fun visitPrimitiveFieldAccess(e: PrimitiveFieldAccess): R
+    fun visitUnaryOperatorExpEmbedding(e: UnaryOperatorExpEmbedding): R
+    fun visitErrorExp(e: ErrorExp): R
+    fun visitGoto(e: Goto): R
+    fun visitInhaleDirect(e: InhaleDirect): R
+    fun visitInhaleInvariants(e: InhaleInvariants): R
+    fun visitNonDeterministically(e: NonDeterministically): R
+    fun visitWhile(e: While): R
+    fun visitFieldAccess(e: FieldAccess): R
+    fun visitInvokeFunctionObject(e: InvokeFunctionObject): R
+    fun visitAssign(e: Assign): R
+    fun visitFieldModification(e: FieldModification): R
+    fun visitLabelExp(e: LabelExp): R
+    fun visitUnitLit(e: UnitLit): R
+    fun visitSharingContext(e: SharingContext): R
+    fun visitWithPosition(e: WithPosition): R
+    fun visitDefault(e: ExpEmbedding): R
+}
+
+internal fun <R> ExpEmbedding.accept(v: ExpVisitor<R>): R = when (this) {
```

Comment:
Please implement this as a virtual function, so that people know they need to add a new case when they add another class.